homeassistant/components/habitica/calendar.py

```diff
@@ -225,3 +229,156 @@ def extra_state_attributes(self) -> dict[str, bool | None] | None:
+
+                if end < start_date:
+                    # Event ends before date range
+                    continue
+
+                if end_date and start > end_date:
+                    # Event starts after date range
+                    continue
+
+                events.append(
+                    CalendarEvent(
+                        start=start,
+                        end=end,
+                        summary=task["text"],
+                        description=task["notes"],
+                        uid=f"{task["id"]}_{reminder["id"]}",
+                    )
+                )
+
+        return sorted(
+            events,
+            key=lambda event: event.start,
+        )
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+        return next(iter(self.reminders(dt_util.now())), None)
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+
+        return self.reminders(start_date, end_date)
+
+
+class HabiticaDailyRemindersCalendarEntity(HabiticaCalendarEntity):
+    """Habitica daily reminders calendar entity."""
+
+    entity_description = CalendarEntityDescription(
+        key=HabiticaCalendar.DAILY_REMINDERS,
+        translation_key=HabiticaCalendar.DAILY_REMINDERS,
+    )
+
+    def start(self, reminder_time: str, reminder_date: date) -> datetime:
+        """Generate reminder times for dailies.
+
+        Reminders for dailies have a datetime but the date part is arbitrary,
+        only the time part is evaluated. The dates for the reminders are the
+        dailies' due dates.
+        """
+        return datetime.combine(
+            reminder_date,
+            datetime.fromisoformat(reminder_time)
+            .replace(
+                second=0,
+                microsecond=0,
+            )
+            .time(),
+            tzinfo=dt_util.DEFAULT_TIME_ZONE,
+        )
+
+    def reminders(
+        self, start_date: datetime, end_date: datetime | None = None
+    ) -> list[CalendarEvent]:
+        """Reminders for dailies."""
+
+        events = []
+        if end_date and end_date < self.today:
+            return []
+        start_date = max(start_date, self.today)
+
+        for task in self.coordinator.data.tasks:
+            if not (task["type"] == HabiticaTaskType.DAILY and task["everyX"]):
+                continue
+
+            recurrences = build_rrule(task)
+            recurrences_start = self.today
+
+            recurrence_dates = self.get_recurrence_dates(
+                recurrences, recurrences_start, end_date
+            )
+            for recurrence in recurrence_dates:
+                is_future_event = recurrence > self.today
+                is_current_event = recurrence <= self.today and not task["completed"]
+
+                if not (is_future_event or is_current_event):
```

Comment:
This is also a little hard to reasona about.  "not future and not current" might be easier to understand? That said, this may be style preference so ok to ignore.

Reply:
changed it here and there's another occurrence above that i would change in the follow-up PR